File: android/app/src/main/java/de/repeatuntil/yata/application/user/UserApplicationService.kt
package de.repeatuntil.yata.application.user

import de.repeatuntil.yata.domain.user.User
import de.repeatuntil.yata.domain.user.UserService

class UserApplicationService(
    private val userService: UserService,
) {

    suspend fun loadOrCreateActiveUser(): Result<User> {
        return runCatching {
            val activeUser = userService.getActiveUser().getOrThrow()
            if (activeUser != null) {
                activeUser
            } else {
                val newUser = User(isActive = true)
                userService.addUser(newUser).getOrThrow()
                newUser
            }
        }
    }
}

File: ios/YATA/Application/User/UserApplicationService.swift
//
//  UserApplicationService.swift
//  YATA
//
//  Created by Alpár Szotyori on 26.05.24.
//

import Foundation

class UserApplicationService {

    let userService: UserService

    init(userService: UserService) {
        self.userService = userService
    }

    func loadOrCreateActiveUser() async -> Result<User, UserApplicationServiceError> {
        switch await userService.getActiveUser() {
        case .success(let user):
            if let user {
                return Result.success(user)
            } else {
                let newUser = User(isActive: true)
                return await userService.addUser(newUser)
                    .map { newUser }
                    .wrapServiceError()
            }
        case .failure(let error):
            return Result.failure(.userServiceError(error))
        }
    }
}

enum UserApplicationServiceError: Error {
    case userServiceError(_ error: UserServiceError)
}

fileprivate extension Result where Failure == UserServiceError {

    func wrapServiceError() -> Result<Success, UserApplicationServiceError> {
        return mapError { error in
            .userServiceError(error)
        }
    }
}
